README.md

```diff
@@ -1 +1,122 @@
 # ChunkedBase.jl
+
+The package handles the ingestion of data chunks and the distribution & synchronization of work that happens on these chunks in parallel. It came into existence while refactoring the `ChunkedCSV.jl` and `ChunkedJSON.jl` packages and was designed to be extended by packages like these. It is a package used to write parser packages.
+
+Specifically, `ChunkedBase.jl` spawns one task which handles IO and behaves as a coordinator, and a configurable number of worker tasks.
+Both CSV and JSONL are textual formats that delimit records by newlines which makes newlines an ideal point to distribute work. One the coordinator task we ingest bytes into a preallocated buffer and then use `NewlineLexers.jl` package to quickly find newlines in it. In turn, these newlines are distributed among worker tasks and the coordinator immediately starts working on a secondary buffer, while the first one is being processed by the workers.
+
+The process looks something like this, with the coordinator task at the bottom:
+
+| ![Diagram](/docs/diagrams/chunked_base.png) |
+|:--:|
+| *The coordinator synchronizes with workers using a counter behind a mutex (there is one per buffer). It splits the newlines into N segments, distributes them, and increments the counter by N. After the coordinator distributes work, it starts to process the second chunk of data, while the first buffer is being worked on. There is a handoff happening between the two buffers -- we need to copy the bytes after the last newline from the first buffer to the second. Each worker decrements after the `consume!` is done, and the coordinator will wait for the counter to reach 0 before it overwrites the buffer with new data.* |
+
+Packages like `ChunkedCSV.jl` and `ChunkedJSON.jl` hook into this structure by defining their own `populate_result_buffer!` methods that parse the records they were assigned into their custom `Result` buffers which are then handed to the `consume!` method (e.g. to be inserted into a database).
```

Comment:
```suggestion
Packages like `ChunkedCSV.jl` and `ChunkedJSON.jl` hook into this structure by defining their own `populate_result_buffer!` methods that parse the records they were assigned into their custom `Result` buffers which are then handed to a custom `consume!` method (e.g. to be inserted into a database).
```
(consume! is also overloaded?)

Reply:
They do e.g. for debugging, and testing but it is RAICode that defines its own `consume!` for CSV, JSONL, and JSONL-general.

Reply:
Oh, interesting. That should definitely be clarified here! Something like ChunkedCSV and ChunkedJSON define their own populate methods, and the _callers_ of ChunkedCSV define their own consume! methods (e.g. ...).

Reply:
I mean it could be defined in ChunkedCSV/ChunkedJSONL but since RAICode want something very specific for itself, then we defined it there. Since this is a public package, I don't think we should mention that some `consume!` methods are defined in RAICode.